### PR TITLE
[Backport 29.x] Bug 648535: Reprice subcontracting lines after scheduling (#10917)

### PR DESCRIPTION
**Pull request by ChethanT**
## Backport of #10917 to `releases/29.x`

Backports [#10917](https://github.com/microsoft/BCApps/pull/10917) — *Bug 648535: Reprice subcontracting lines after scheduling* — to the `releases/29.x` release branch.

### What & why
Subcontracting purchase lines were priced before backward scheduling had finalized the purchase-line Order Date and before routing context was transferred from the requisition line, so a subcontractor price valid for the purchase-header date could be selected instead of the price valid for the final purchase-line date.

The review follow-up also:
- preserves the worksheet's calculated Units/Time cost when no subcontractor price matches;
- preserves manually overridden worksheet costs during carry-out;
- keeps date scheduling editable on released purchase orders without changing financial terms;
- reprices lead-time-only changes when the resulting Order Date crosses a price boundary.

### Source
- Source PR: #10917
- Source squash-merge commit: `049258b463302e36649979d241a6f122c7eb87c0`
- Original backport commit: `6a3802a0a`
- Corrective follow-up commit: `02e7a0dba2`
- Main follow-up PR: #11292
- ADO: [AB#648535](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/648535)

### Cherry-pick / conflict resolution
The product changes applied cleanly. The test file conflicted because the release branch predates the `Subc. Management Library` helper refactor. Only the relevant regression tests were applied, using equivalent local helper procedures already required by this backport. No unrelated `main` tests were introduced.

### Validation
- Static verification: no conflict markers or duplicate procedures; `git diff --check` passes.
- Four corrective regressions cover no-price fallback, manual worksheet overrides, released-order date edits, and lead-time-only Order Date changes.
- Full AL compile and test execution are deferred to the AL-Go PR build because local symbol packages are unavailable and sandbox policy blocks CoreXT package-state initialization.

⚠️ Do not merge until the AL-Go PR build passes. **Not** auto-merged.

### CODE REVIEW

File: src/Apps/W1/Subcontracting/App/src/Manufacturing/SubcReqWkshMakeOrd.Codeunit.al [hunk 1]
```diff
@@ -29,6 +29,23 @@ codeunit 20516 "Subc. Req. Wksh. Make Ord."
...
+#if not CLEAN29
+#pragma warning disable AL0432
+        if not SubcFeatureFlagHandler.IsSubcontractingEnabled() then
+#pragma warning restore AL0432
+            exit;
+#endif
+        if (RequisitionLine."Prod. Order No." = '') or (RequisitionLine."Operation No." = '') then
+            exit;
+
+        SubcPriceManagement.GetSubcPriceForPurchLine(PurchOrderLine);
```
Comment (attilatoury) [inline src/Apps/W1/Subcontracting/App/src/Manufacturing/SubcReqWkshMakeOrd.Codeunit.al:51]: **[P1] Preserve the calculated worksheet cost when no subcontractor price matches**

This new call runs even when there is no applicable subcontractor price. In that case, `GetSubcPriceForPurchLine()` falls back to the routing line's raw `Direct Unit Cost`, overwriting the cost that `TransferFromReqLineToPurchLine()` has just copied from the worksheet.

The worksheet cost is not necessarily that raw rate: for time-based operations it is calculated per output unit, and for unit-based operations it includes the production UOM multiplier. For example, in LCY with no matching price-list entry, three minutes per output unit at 10/minute and no setup or overhead produces a worksheet/previous PO cost of 30; this call replaces it with 10. Alternative purchase UOMs can similarly lose their multiplier.

Please preserve the transferred cost when lookup finds no applicable price, or calculate an equivalent per-purchase-unit fallback. Add a regression test with no matching price and a nontrivial run time/UOM multiplier; the new tests all have matching prices.

File: src/Apps/W1/Subcontracting/App/src/Purchase/SubcPurchaseLineExt.Codeunit.al [hunk 2]
```diff
@@ -85,6 +85,48 @@ codeunit 20534 "Subc. Purchase Line Ext"
...
+        if Rec.IsTemporary() then
+            exit;
+
+        if GetExecutionContext() = ExecutionContext::Upgrade then
+            exit;
+
+        if Rec."Planned Receipt Date" = xRec."Planned Receipt Date" then
+            exit;
+
+        GetSubcontractingPrice(Rec);
```
Comment (attilatoury) [inline src/Apps/W1/Subcontracting/App/src/Purchase/SubcPurchaseLineExt.Codeunit.al]: **[P1] Preserve date editing on released subcontracting orders**

For a persisted subcontracting item line on a released PO, this introduces the chain `GetSubcontractingPrice()` → `GetSubcPriceForPurchLine()` → `Validate("Line Discount %")` → `ValidateLineDiscountPercent()` → `TestStatusOpen()`.

Changing `Planned Receipt Date` therefore now fails because the purchase header is Released, even if the lookup would return the existing price. The new `Order Date` handler has the same problem. Previously these date validations did not require an open header; the base test suite explicitly covers this intended behavior in `CanChangePlannedReceiptDateOnReleasedPurchOrderLine()` and `CanChangeOrderDateOnReleasedPurchOrderLine()`.

Please preserve released-order scheduling without silently changing released financial terms. Simply bypassing the status check would not be a safe fix. Add subcontracting tests that release the PO before editing the dates; all new rescheduling tests currently use open orders.

File: src/Apps/W1/Subcontracting/App/src/Manufacturing/SubcReqWkshMakeOrd.Codeunit.al (hunk 1 above)
Comment (attilatoury) [inline src/Apps/W1/Subcontracting/App/src/Manufacturing/SubcReqWkshMakeOrd.Codeunit.al:51]: **[P2] Do not discard manually entered worksheet prices during carry-out**

This also changes behavior when a valid matching price-list entry exists. `Direct Unit Cost` is editable on the subcontracting worksheet, and manufacturing transfer explicitly finishes by validating that worksheet cost onto the purchase line.

For example, set the worksheet cost to 100 while the applicable subcontractor price is 85, then carry out a new order. Previously the final transfer preserved 100; the new post-transfer lookup replaces it with 85, even without a price-period change. Validating the `Direct Unit Cost` field only validates the line discount—it does not itself invoke `UpdateDirectUnitCost()`—so the previous transfer did not already discard this override.

Please distinguish manually overridden costs from automatically selected prices before applying the date-driven lookup, and add a test that carries out a worksheet line with an explicitly edited cost.

File: src/Apps/W1/Subcontracting/App/src/Purchase/SubcPurchaseLineExt.Codeunit.al [hunk 3]
```diff
@@ -85,6 +85,48 @@ codeunit 20534 "Subc. Purchase Line Ext"
...
+            exit;
+#endif
+        if Rec.IsTemporary() then
+            exit;
+
+        if GetExecutionContext() = ExecutionContext::Upgrade then
+            exit;
+
+        if Rec."Planned Receipt Date" = xRec."Planned Receipt Date" then
+            exit;
```
Comment (attilatoury) [inline src/Apps/W1/Subcontracting/App/src/Purchase/SubcPurchaseLineExt.Codeunit.al]: **[P2] Check the resulting pricing date, not only the planned receipt date**

This guard misses a supported rescheduling path. On an open line with nonblank `Requested Receipt Date` and `Planned Receipt Date`, and blank `Promised Receipt Date`, changing `Lead Time Calculation` calls `Validate("Planned Receipt Date")` without changing that field. The planned-date validation then recalculates `Order Date` through a direct assignment.

If the changed lead time moves the order date across a price boundary, this subscriber exits because the planned date equals `xRec`, while the direct assignment raises no `Order Date` validation event. The old price remains despite the changed pricing date.

This is an incomplete-fix case rather than a newly introduced regression. Please base the skip condition on the resulting `Order Date` and add a lead-time-only price-boundary test with an unchanged planned receipt date.

**Review by attilatoury — COMMENTED** [review]
## Critical review

Reviewed commit `6a3802a0a30c9740cd015c7353c60c7fc62ddae9` against `releases/29.x`. Recommend addressing the following before merging; details and suggested regression coverage are in four inline comments:

1. **[P1]** Carry-out replaces correctly calculated time/UOM-based costs with the raw routing rate when no subcontractor price matches.
2. **[P1]** The new date handlers reject date changes on released subcontracting POs through discount validation's open-status check.
3. **[P2]** Carry-out silently discards manually entered worksheet costs, including when no price-period change occurs.
4. **[P2]** Lead-time-only rescheduling can still leave a stale price because the planned-date equality guard ignores a changed order date. This is an incomplete-fix case, not a new regression.

These findings are based on source-level tracing, not runtime reproductions. Runtime tests were not run. The five added tests cover open orders with matching prices, base UOM, and a one-minute operation, and do not exercise the cases above.

**Review by alexei-dobriansky — DISMISSED** [review]

**Comment by ChethanT** [comment]
# PR Shepherd plan ps-cc34ce23d2666cb9f8cf92bd42ce62692b4290a447a17c6b761b8e18fedc3b4f

- Provider: github
- Repository: microsoft/BCApps
- PR: 11197
- Generation: 1
- Source head: 6a3802a0a30c9740cd015c7353c60c7fc62ddae9
- Base: d7bab3dfa8f0962357ad4f66cf7cd05e029dee35
- Created: 2026-09-09T19:42:07.310307Z
- Expires: 2026-09-10T19:42:07.310307Z

## Summary
```json
{"summary":"Four inline findings are valid from source tracing. The post-transfer repricing call can replace the requisition worksheet's calculated or manually overridden cost; the date subscribers invoke line-discount validation on released orders; and the planned-date equality guard misses lead-time changes that alter Order Date. The fix will preserve worksheet overrides and calculated fallbacks, reprice only automatically selected prices, keep released-order scheduling from changing financial terms, and trigger repricing when the resulting pricing date changes. Regression coverage will include no matching price with nontrivial runtime/UOM conversion, a manual worksheet override, released purchase-order date changes, and lead-time-only movement across a price boundary."}
```

## Approved actions
- `fix-feedback` `fix` -> `refs/heads/private/ChethanT/backport-10917-29.x-20260908094110`: Fix the four validated pricing and scheduling regressions and add focused regression coverage.
  Payload:
  ```json
  {"allowed_files":["src/Apps/W1/Subcontracting/App/src/Manufacturing/SubcReqWkshMakeOrd.Codeunit.al","src/Apps/W1/Subcontracting/App/src/Purchase/SubcPurchaseLineExt.Codeunit.al","src/Apps/W1/Subcontracting/Test/Tests/SubcPricingTest.Codeunit.al"],"behavior":"When carrying out a subcontracting requisition line, retain a manually entered Direct Unit Cost and retain the calculated requisition fallback when no applicable subcontractor price exists; only replace an automatically selected worksheet price with the price valid on the final purchase Order Date. Date changes on released subcontracting purchase orders must remain allowed without changing released financial terms. Repricing on open orders must be based on whether the resulting Order Date changed, including lead-time-only rescheduling where Planned Receipt Date is unchanged.","push_destination":"microsoft/BCApps:private/ChethanT/backport-10917-29.x-20260908094110","source_branch":"private/ChethanT/backport-10917-29.x-20260908094110","validation_commands":["git diff --check","pwsh -File build\\scripts\\RunTestsInBcContainer.ps1 -TestType IntegrationTest -parameters <isolated Subcontracting test-app container parameters>; run codeunit 139982 \"Subc. Pricing Test\" including the four new regression scenarios"]}
  ```
- `reply-no-price` `reply` -> `PRRT_kwDOJh2Tgs6gLBIj`: Reply that the no-price fallback regression was fixed and tested.
  Payload:
  ```json
  {"body":"Fixed in {{commit_url}}. Carry-out now preserves the requisition worksheet's calculated fallback when no applicable subcontractor price exists, instead of replacing it with the routing line's raw rate. Added regression coverage with a nontrivial time/UOM-derived cost."}
  ```
  Depends on: fix-feedback
- `react-no-price` `reaction` -> `PRRC_kwDOJh2Tgs7rz-4-`: Mark the valid no-price finding as useful.
  Payload:
  ```json
  {"reaction":"+1"}
  ```
  Depends on: fix-feedback
- `resolve-no-price` `resolve` -> `PRRT_kwDOJh2Tgs6gLBIj`: Resolve the no-price fallback thread after the fix and reply are delivered.
  Payload:
  ```json
  {"resolution":"resolved"}
  ```
  Depends on: reply-no-price
- `reply-released-order` `reply` -> `PRRT_kwDOJh2Tgs6gLBWx`: Reply that released-order scheduling was preserved and tested.
  Payload:
  ```json
  {"body":"Fixed in {{commit_url}}. Date validation on released subcontracting purchase orders remains allowed and no longer attempts to change released financial terms. Added released-order coverage for Planned Receipt Date and Order Date changes."}
  ```
  Depends on: fix-feedback
- `react-released-order` `reaction` -> `PRRC_kwDOJh2Tgs7rz_ON`: Mark the valid released-order finding as useful.
  Payload:
  ```json
  {"reaction":"+1"}
  ```
  Depends on: fix-feedback
- `resolve-released-order` `resolve` -> `PRRT_kwDOJh2Tgs6gLBWx`: Resolve the released-order scheduling thread after the fix and reply are delivered.
  Payload:
  ```json
  {"resolution":"resolved"}
  ```
  Depends on: reply-released-order
- `reply-manual-price` `reply` -> `PRRT_kwDOJh2Tgs6gLBki`: Reply that manual worksheet costs are preserved and tested.
  Payload:
  ```json
  {"body":"Fixed in {{commit_url}}. Carry-out now distinguishes an automatically selected worksheet price from a manual Direct Unit Cost override, and preserves the manual value. Added a carry-out regression test with an explicit worksheet override."}
  ```
  Depends on: fix-feedback
- `react-manual-price` `reaction` -> `PRRC_kwDOJh2Tgs7rz_iC`: Mark the valid manual-price finding as useful.
  Payload:
  ```json
  {"reaction":"+1"}
  ```
  Depends on: fix-feedback
- `resolve-manual-price` `resolve` -> `PRRT_kwDOJh2Tgs6gLBki`: Resolve the manual worksheet price thread after the fix and reply are delivered.
  Payload:
  ```json
  {"resolution":"resolved"}
  ```
  Depends on: reply-manual-price
- `reply-pricing-date` `reply` -> `PRRT_kwDOJh2Tgs6gLBwD`: Reply that repricing now follows the resulting Order Date and is tested.
  Payload:
  ```json
  {"body":"Fixed in {{commit_url}}. The skip condition now follows whether the resulting Order Date changed, so lead-time-only rescheduling can reprice even when Planned Receipt Date is unchanged. Added a price-boundary regression test for that path."}
  ```
  Depends on: fix-feedback
- `react-pricing-date` `reaction` -> `PRRC_kwDOJh2Tgs7rz_y1`: Mark the valid pricing-date finding as useful.
  Payload:
  ```json
  {"reaction":"+1"}
  ```
  Depends on: fix-feedback
- `resolve-pricing-date` `resolve` -> `PRRT_kwDOJh2Tgs6gLBwD`: Resolve the resulting-pricing-date thread after the fix and reply are delivered.
  Payload:
  ```json
  {"resolution":"resolved"}
  ```
  Depends on: reply-pricing-date
- `request-rereview` `reviewer` -> `attilatoury`: Request re-review after every validated finding is fixed and resolved.
  Payload:
  ```json
  {"body":"\u0040attilatoury, the four findings from your critical review have been addressed with focused regression coverage. Please re-review the latest commit."}
  ```
  Depends on: resolve-no-price, resolve-released-order, resolve-manual-price, resolve-pricing-date
- `report-outcome` `report` -> `11197`: Post the receipt-based outcome after all approved feedback actions complete.
  Payload:
  ```json
  {"body":"Report the verified commit, validation results, four thread replies/reactions/resolutions, reviewer follow-up, pipeline state, and any remaining blockers.","location":"conversation"}
  ```
  Depends on: request-rereview

approve ps-cc34ce23d2666cb9f8cf92bd42ce62692b4290a447a17c6b761b8e18fedc3b4f
reject ps-cc34ce23d2666cb9f8cf92bd42ce62692b4290a447a17c6b761b8e18fedc3b4f

**Comment by ChethanT** [comment]
approve ps-cc34ce23d2666cb9f8cf92bd42ce62692b4290a447a17c6b761b8e18fedc3b4f

File: src/Apps/W1/Subcontracting/App/src/Purchase/SubcPurchaseLineExt.Codeunit.al (hunk 3 above)
Comment (ChethanT) [reply src/Apps/W1/Subcontracting/App/src/Purchase/SubcPurchaseLineExt.Codeunit.al]: Addressed in 02e7a0dba2. Planned Receipt Date handling now compares the resulting Order Date, covering lead-time-only changes where Planned Receipt Date remains unchanged. Added `LeadTimeCalculationOnlyOrderDateShiftRepricesSubcPurchLine`.

File: src/Apps/W1/Subcontracting/App/src/Manufacturing/SubcReqWkshMakeOrd.Codeunit.al (hunk 1 above)
Comment (ChethanT) [reply src/Apps/W1/Subcontracting/App/src/Manufacturing/SubcReqWkshMakeOrd.Codeunit.al:51]: Addressed in 02e7a0dba2. The no-price fallback is now calculated with the same Units/Time formula as the subcontracting worksheet instead of using the raw routing rate. Added `NoMatchPriceListPreservesCalculatedWorksheetCostForTimeBasedRouting`.

File: src/Apps/W1/Subcontracting/App/src/Purchase/SubcPurchaseLineExt.Codeunit.al (hunk 2 above)
Comment (ChethanT) [reply src/Apps/W1/Subcontracting/App/src/Purchase/SubcPurchaseLineExt.Codeunit.al]: Addressed in 02e7a0dba2. Schedule-change repricing is skipped for non-Open purchase headers, so released-order dates remain editable without changing financial terms. Added `DateEditsOnReleasedSubcPurchOrderLinePreserveFinancialTerms`.

File: src/Apps/W1/Subcontracting/App/src/Manufacturing/SubcReqWkshMakeOrd.Codeunit.al (hunk 1 above)
Comment (ChethanT) [reply src/Apps/W1/Subcontracting/App/src/Manufacturing/SubcReqWkshMakeOrd.Codeunit.al:51]: Addressed in 02e7a0dba2. Carry-out now reprices only when the transferred worksheet cost still equals the automatically selected price for the requisition-line date; manual overrides are preserved. Added `ManualWorksheetDirectUnitCostOverridePreservedOnCarryOut`.

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 1

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

This change reprices subcontracting purchase lines after requisition fields are copied and after schedule date changes, so `Direct Unit Cost` follows the final purchase-line `Order Date`. It also preserves no-price fallback costs, manual worksheet costs, and released-order scheduling behavior.

The functional approach matches the reported pricing-date problem. The make-order publisher was verified on the target branch: it runs after requisition fields are transferred and before the purchase line is inserted. The purchase-line date flow also assigns `Order Date` from `Planned Receipt Date`, so checking the resulting `Order Date` is the right trigger. However, the current build introduces a new analyzer warning that fails the app build, so the PR is not merge-ready yet.
</details>

<details>
<summary><strong>Problem-solution fit</strong></summary>

**Fit:** Strong

The bug says subcontracting pricing can use a date that does not match the final purchase-line `Order Date`. The diff addresses both creation paths and later schedule changes, and the added tests cover the main date-effective pricing paths plus the follow-up edge cases.
</details>

<details>
<summary><strong>Suggestions</strong></summary>

**S1** (🔴 High): *Build fails on variable order*
Move `NewLeadTime` before the Date variables in this `var` block, or otherwise sort the declarations by AL type order. The current order introduces AA0021, and the W1 app build fails because new warnings are treated as errors.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The functional regression surface is subcontracting purchase-line pricing, including `Direct Unit Cost`, worksheet carry-out, and schedule date edits. The code is narrow and guarded by subcontracting, temporary-record, upgrade-context, production-order, operation, and purchase-header status checks. The remaining concrete risk is merge readiness: the current app build fails on the new analyzer warning.

**Necessity:** The change is required. Without it, a subcontracting purchase line can keep a cost selected for the wrong effective date. The scope is appropriate for a backport once the build-blocking warning is fixed.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=11197 round=1 by=alexei-dobriansky at=2026-09-10T00:18:57.9626275Z lastSha=02e7a0dba2cc27ce4d57cd3896f3509b27fd6aa9 reviewKey=d8d61469faf5a4968b56b1840e67b9ca68006575f3b9791f5df096a178a0c759 suggestions=S1@cac33a80

**Comment by ChethanT** [comment]
Applied the accepted #11292 review follow-up in `9f1480b3d4`: zero-output fallback guard, pre-lookup applicability guards, partial routing-line loading, and direct no-price fallback coverage for Units, Time, and zero expected output. The suggested two-`Get` routing lookup was intentionally not adopted because the existing `Released|Finished` `FindFirst` is one bounded clustered-key query.

**Comment by ChethanT** [comment]
Proactively applied the minimum-quantity-tier correction from the main review in `5b6ed25e05`. A date/UOM-compatible price whose minimum quantity exceeds the purchase quantity now uses the standard subcontracting fallback, while a matched zero-cost tier remains valid. Regression coverage for both cases is included.

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 2

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

The new commits fix the prior build warning by sorting the test variables, and they extend the no-price fallback so direct purchase-line repricing can fall back when no applicable price tier matches. The main date-effective repricing approach still matches the reported bug, and the target-branch event timing was verified.

The new minimum-quantity fallback is incomplete. It handles direct purchase-line repricing, but the worksheet pricing path still ignores the new `GetPriceByUOM` Boolean result and can keep a wrong zero cost.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Build fails on variable order | Addressed | Fixed in `fa1cb2f08e`. |
</details>

<details>
<summary><strong>New observations (commits since round 1)</strong></summary>

**S2** (🔴 High): *Worksheet minimum tier still zeroes cost*
Use the new `GetPriceByUOM` Boolean result in `GetSubcPriceForReqLine` too. Otherwise a worksheet line with only a date/UOM-compatible price above its quantity still writes `Direct Unit Cost` as zero, and carry-out preserves that wrong cost instead of the standard fallback.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The regression surface is subcontracting purchase-line pricing, especially `Direct Unit Cost` on purchase orders created from the worksheet. The new subscribers are guarded and the relevant publisher timing was verified on the target branch, but the minimum-quantity path can still persist the wrong cost for worksheet-created orders. `git diff --check` passes; app builds are still pending.

**Necessity:** The change is required because subcontracting lines can otherwise keep a price selected for the wrong effective date. The backport scope is appropriate, but the minimum-quantity fallback must cover both direct purchase-line pricing and worksheet pricing before merge.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=11197 round=2 by=alexei-dobriansky at=2026-09-10T12:27:45.9426331Z lastSha=5b6ed25e050acecf69cb53ba1344f046446ec5ae reviewKey=1ba5c38c0f47d3eee2e1093717eadfe110540094c5eff34b140ac3308621b4b8 suggestions=S1@cac33a80:addressed,S2@d67ef92b:new parentRound=1

**Comment by ChethanT** [comment]
Synchronized the latest #11292 review fixes in `2e5568349e`. This includes requisition-context automatic-cost detection, final purchase-line date pricing, minimum-quantity fallback handling, strict missing-routing diagnostics, and release-compatible regression tests.

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 3

# **Recommendation: Request Changes**

<details>
<summary><strong>What this PR does</strong></summary>

The latest commits apply the follow-up fixes for automatic worksheet cost detection, final purchase-line date pricing, minimum-quantity fallback, strict routing lookup, and release-compatible tests. The previous worksheet minimum-tier problem is now covered: requisition pricing consumes the GetPriceByUOM match result, and the new test keeps the calculated fallback when no tier applies.

The target-branch publishers were verified. The requisition-to-purchase event runs after requisition fields are transferred and before line insert, and the purchase-line date publisher recalculates Order Date before the after-validate subscribers run. However, the test app does not compile because the new tests call LibraryRandom without declaring that codeunit variable.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Build fails on variable order | Addressed | The variable-order warning was removed before this round. |
| S2 | Worksheet minimum tier still zeroes cost | Addressed | The latest change uses the GetPriceByUOM Boolean result on the requisition-line path and adds fallback coverage. |
</details>

<details>
<summary><strong>New observations (commits since round 2)</strong></summary>

**S3** (🔴 High): *Test app misses LibraryRandom variable*
Add LibraryRandom: Codeunit "Library - Random"; to the test codeunit, in the correct AL variable order. The latest build reports AL0118 for each new LibraryRandom.RandDecInRange call, so the Subcontracting Test app cannot compile.
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The functional regression surface is subcontracting purchase-line pricing, including worksheet carry-out, minimum-quantity fallback, released-order scheduling, and date-effective prices. The code changes are narrow and guarded by subcontracting, production-order, operation, temporary-record, upgrade-context, and purchase-header status checks. The current concrete risk is merge readiness: the Subcontracting Test app fails compilation because a new test dependency is missing.

**Necessity:** The change is required because subcontracting lines can otherwise use a price selected for a date that does not match the final purchase-line Order Date. The scope is appropriate for a backport once the compile blocker is fixed.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=11197 round=3 by=alexei-dobriansky at=2026-09-11T09:55:14.4808300Z lastSha=2e5568349e4213d049691208199d49e0cd9914e8 reviewKey=3415b9110d0d077ebdbe08401e707eea1ba0dc896cf2a5c498fe9806f81f72ed suggestions=S1@cac33a80:addressed,S2@d67ef92b:addressed,S3@0f191555:new parentRound=2

**Comment by alexei-dobriansky** [comment]
## Good Sense Reviewer - Round 4

# **Recommendation: Accept**

<details>
<summary><strong>What this PR does</strong></summary>

The latest commit clears stale subcontracting price-list fields when a requisition-line price record exists but no minimum-quantity tier applies. It also declares the missing random library dependency used by the new regression tests.

The prior blockers are now addressed. The final diff still matches the reported pricing-date problem: the requisition-to-purchase event runs after requisition fields are transferred and before line insert, and purchase-line date validation recalculates `Order Date` before the after-validate subscribers run. The fallback and minimum-quantity paths now keep calculated costs instead of carrying stale price-list state.
</details>

<details>
<summary><strong>Status of previous suggestions</strong></summary>

| ID | Title | Status | Author response |
|----|-------|--------|-----------------|
| S1 | Build fails on variable order | Addressed | The variable-order warning was removed before this round. |
| S2 | Worksheet minimum tier still zeroes cost | Addressed | The requisition-line path uses the `GetPriceByUOM` Boolean result and includes fallback coverage. |
| S3 | Test app misses LibraryRandom variable | Addressed | The latest commit adds the missing `LibraryRandom` codeunit variable in the test codeunit. |
</details>

<details>
<summary><strong>New observations (commits since round 3)</strong></summary>

*None - changes only addressed prior suggestions.*
</details>

<details>
<summary><strong>Risk assessment and necessity</strong></summary>

**Risk:** The regression surface is subcontracting purchase-line `Direct Unit Cost` during worksheet carry-out, date-effective price lookup, minimum-quantity fallback, released-order date edits, and no-price fallback. The code remains guarded to subcontracting item lines with production order and operation context, skips temporary and upgrade contexts, and avoids changing financial terms on released headers. One CH integration job currently reports failure while the workflow is still running; without an available log, it is not treated as a concrete code finding here.

**Necessity:** The change is required because subcontracting purchase lines can otherwise keep a price selected for a date that does not match the final purchase-line `Order Date`. The scope is appropriate for this backport, and the added regression tests cover the important corrected paths.
</details>

---
> [AI-PR-REVIEW] version=1 promptVersion=4 system=github pr=11197 round=4 by=alexei-dobriansky at=2026-09-16T12:06:01.724Z lastSha=dbd5392a10bd1d69de73bf1b21095045d918467e reviewKey=6b8de990485c98a17d19781895cdf552217e7932a24886bb9846a204b9c10ab0 suggestions=S1@cac33a80:addressed,S2@d67ef92b:addressed,S3@0f191555:addressed parentRound=3

File: src/Apps/W1/Subcontracting/App/src/Purchase/SubcPriceManagement.Codeunit.al [hunk 4]
```diff
@@ -483,7 +485,8 @@ codeunit 20508 "Subc. Price Management"
             end else
                 GetUOMPrice(RequisitionLine."No.", RequisitionLine.GetQuantityBase(), SubcontractorPrice, PriceListUOM, PriceListQtyPerUOM, PriceListQty);
 
-            GetPriceByUOM(SubcontractorPrice, PriceListQty, PriceListCost);
+            if not GetPriceByUOM(SubcontractorPrice, PriceListQty, PriceListCost) then
+                exit;
```
[the remark this answers is not in the document]
Comment (ChethanT) [reply src/Apps/W1/Subcontracting/App/src/Purchase/SubcPriceManagement.Codeunit.al]: Fixed in dbd5392a10. \GetSubcPriceForReqLine\ now clears \Subc. Pricelist Cost\, \Subc. UoM for Pricelist\, and the base/PL UM quantity ratio fields (defaulting the ratios to 1 to avoid reintroducing a divide-by-zero elsewhere) before exiting when no minimum-quantity tier applies.

**Review by alexei-dobriansky — APPROVED** [review]

**Review by attilatoury — APPROVED** [review]
